### PR TITLE
[R-package] fixed sorting issues in lgb.cv() when using a model with observation weights (fixes #2572)

### LINKED ISSUE #2572: [R-package] sorting of observation weights and initial scores is broken in lgb.cv()
In #2503 , it was reported that using certain sampling strategies could cause incorrect results if the indices for subsetting weren't sorted. This was "fixed" in the R package by #2524 , however that PR introduced a bug noted in https://github.com/microsoft/LightGBM/pull/2524#issuecomment-545719165.

Essentially, `slice.lgb.Dataset()` takes in a vector of indices (row numbers) and returns a new `Dataset` with that subset. On its way to doing that, it sorts those indices. If a model had observation weights or initial scores, they don't get sorted the same way, meaning they'll refer to the wrong observations.

To fix this, need to ensure that tuples of indices, weights, and initial scores are all kept together when re-sorting indices.

### PR DESCRIPTION
**Pull request by jameslamb**
See #2572 for details.

### CODE REVIEW

**Comment by StrikerRUS** [comment]
Please rebase to the latest `master` to spot some linting issues.

File: R-package/R/lgb.cv.R [hunk 1]
```diff
@@ -267,35 +267,56 @@ lgb.cv <- function(params = list()
...
+        group <- folds[[k]]$group
+      } else {
+        test_indices <- folds[[k]]
+        group <- NULL
+      }
+      train_indices <- seq_len(nrow(data))[-test_indices]
+
+      # set up test set
+      indexDT <- data.table::data.table(
+        indices = train_indices
```
Comment (StrikerRUS) [inline R-package/R/lgb.cv.R]: `train_indices` to set up "test set"?

File: R-package/R/lgb.cv.R [hunk 2]
```diff
@@ -267,35 +267,56 @@ lgb.cv <- function(params = list()
...
+        , init_score = getinfo(data, "init_score")[train_indices]
+      )
+      data.table::setorderv(indexDT, "indices", order = 1L)
+      dtrain <- slice(data, indexDT$indices)
+      setinfo(dtrain, "weight", indexDT$weight)
+      setinfo(dtrain, "init_score", indexDT$init_score)
+
+      if (!is.null(group)) {
+        setinfo(dtest, "group", group)
+        setinfo(dtrain, "group", group)
```
Comment (StrikerRUS) [inline R-package/R/lgb.cv.R]: Is `group` fields shared across train and test sets?

**Comment by jameslamb** [comment]
> Sorry if my comments are not so smart, but since review was requested from me, let me show which things look confusing from point of view of first-time code reader.

thanks for the review! I really appreciate it. They are good comments 😀

File: R-package/R/lgb.cv.R (hunk 2 above)
Comment (jameslamb) [reply R-package/R/lgb.cv.R]: it was in the original code. Let me double-check what that is doing.

File: R-package/R/lgb.cv.R (hunk 1 above)
Comment (jameslamb) [reply R-package/R/lgb.cv.R]: ah! there is a lot of repetition in this code right now between train and test sets, so this was bound to happen. Let me think if there is a cleaner way to do this that reduces the duplication...I think I have an idea how

File: R-package/R/lgb.cv.R (hunk 2 above)
Comment (jameslamb) [reply R-package/R/lgb.cv.R]: ok I see now that I misunderstood how `group` is working. I haven't worked with the learning-to-rank side of `LightGBM` at all, and I'm struggling to understand [this documentation](https://github.com/microsoft/LightGBM/blob/master/docs/Parameters.rst#query-data).

It seems that `group` going to a vector like `c(10,  15, 20)` which means "the first 10 records are from one query, the next 15 are samples from another query, etc.".

If that's true, I don't understand what `getinfo(data, "group")[-folds[[k]]$group]` could be doing. We don't have any R learning-to-rank examples I could find, so I need some help from @guolinke or someone else.

**Comment by guolinke** [comment]
Very sorry for the late response, I will take a look today

File: R-package/R/lgb.cv.R (hunk 2 above)
Comment (guolinke) [reply R-package/R/lgb.cv.R]: please refer to the python implementation: https://github.com/microsoft/LightGBM/blob/master/python-package/lightgbm/engine.py#L308-L324

Comment (guolinke) [reply R-package/R/lgb.cv.R]: I think there is a similar logics in R: https://github.com/microsoft/LightGBM/blob/master/R-package/R/lgb.cv.R#L409-L438

Comment (jameslamb) [reply R-package/R/lgb.cv.R]: Sorry, I have been staring at this code for a while and am still confused what it's doing. I am going to try to figure out a learning-to-rank example in R (since we don't have any in the documentation) and use that to step through the code. Going to try to figure out how to do this from the Python tests using lambdarank: https://github.com/microsoft/LightGBM/blob/master/tests/python_package_test/test_engine.py#L635

Comment (guolinke) [reply R-package/R/lgb.cv.R]: I remember this is written by @Laurae2

Comment (jameslamb) [reply R-package/R/lgb.cv.R]: I think this is working! Thanks to @Laurae2 for pointing me to the examples of learning-to-rank code for the R package in #791 and #397. Based on those, I've added unit tests on `lgb.train()` and `lgb.cv()` that at least cover the LTR-specific branches of that code.

Reviewers, could you take another look at this PR when you have a chance?

Comment (jameslamb) [reply R-package/R/lgb.cv.R]: oh also I did rebase to most recent `master`, to get the recent minor changes to the R package included (#2652 and #2654) in the tests

File: R-package/tests/testthat/test_learning_to_rank.R [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+    )
+    nfold <- 4L
+    model <- lgb.cv(
+        params = params
+        , data = dtrain
+        , nrounds = 10L
+        , nfold = nfold
+        , min_data = 1L
+        , learning_rate = 1.0
+        , early_stopping_rounds = 10L
```
[the remark this answers is not in the document]
Comment (jameslamb) [reply R-package/tests/testthat/test_learning_to_rank.R]: 😬 bad copy-pasting, thank you

File: R-package/tests/testthat/test_learning_to_rank.R [hunk 4]
```diff
@@ -0,0 +1,94 @@
+context("Learning to rank")
```
[the remark this answers is not in the document]
Comment (jameslamb) [reply R-package/tests/testthat/test_learning_to_rank.R]: totally agree. Since this is currently a user-facing bug (i.e. the functionality is already incorrect on `master`, #2572 ), I wanted to get something that I think is a fix up as fast as possible. I'm happy to keep expanding these tests and pushing on this PR until @Laurae2 has a chance to review the changes to `lgb.cv()`, but once we get that approval I don't think we should delay merging this for the sake of more testing.

[the remark this answers is not in the document]
Comment (StrikerRUS) [reply R-package/tests/testthat/test_learning_to_rank.R]: Sure! I think adding a check for adequate values of NDCG will be enough for now.

File: R-package/tests/testthat/test_learning_to_rank.R (hunk 3 above)
[the remark this answers is not in the document]
Comment (jameslamb) [reply R-package/tests/testthat/test_learning_to_rank.R]: I decided to drop early stopping altogether in these tests

File: R-package/tests/testthat/test_learning_to_rank.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (jameslamb) [reply R-package/tests/testthat/test_learning_to_rank.R]: Thanks for the  feedback @StrikerRUS , very good idea. I've added more to these tests in https://github.com/microsoft/LightGBM/pull/2573/commits/f14458bd3678d8b8883f8a8c38a750c769b8f223

I'm testing the  following things for  learning-to-rank with  `lambdarank`.  I think this should be sufficient.

`lgb.train()`

* returns an `lgb.Booster`  object, with the following properties
    - `max_feature_idx` matching dimensions of the training data
    -  one set  of `eval_results` per metric (3 in these tests, since I had  `ndcg_at = "1,2,3"`
     - all evaluation results have reasonable values (between 0.0 and 1.0 since they're nDCG)

`lgb.cv()`

* returns an `lgb.CVBooster` object, with the following properties
    - `boosters`, containing a list of length `nfold`  of  `lgb.Booster`
    - `best_iter`  with a reasonable value  (>1, <=  `nrounds`)
    - `best_score`  is based on the first metric (`ncdg@1`, in this case)
    -  one set  of `eval_results` per metric (3 in these tests, since I had  `ndcg_at = "1,2,3"`. For which:
         - `eval` and `eval_err`  are arrays with length `nrounds`  (since  I did not use  early stopping)
          - all `eval`  values  have reasonable values (between 0.0 and 1.0 since they're nDCG)


@guolinke or  @Laurae2 could I get one more review of the changes in this PR from one of you?

**Comment by jameslamb** [comment]
I see there are some lingering linting issues. Will fix shortly!

File: R-package/tests/testthat/test_learning_to_rank.R [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+test_that("learning-to-rank with lgb.cv() works as expected", {
+    data(agaricus.train, package = "lightgbm")
+    # just keep a few features,to generate an model with imperfect fit
+    train <- agaricus.train
+    train_data <- train$data[, 1:20]
+    dtrain <- lgb.Dataset(
+        train_data
+        , label = train$label
+        , group = c(rep(4L, 1500L), 513L)
```
Comment (Laurae2) [inline R-package/tests/testthat/test_learning_to_rank.R]: Should there be 1501 groups? (1500 groups of length 4 and 1 group of length 513?)

File: R-package/tests/testthat/test_learning_to_rank.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (StrikerRUS) [reply R-package/tests/testthat/test_learning_to_rank.R]: @jameslamb 

> all evaluation results have reasonable values (between 0.0 and 1.0 since they're nDCG)

I think that tests can be more strict to the final metric value in that aspect to ensure that training actually happens. Refer to Python tests as an example:
https://github.com/microsoft/LightGBM/blob/301402c83bc732f4ec41135cc5e5f6f5984348d8/tests/python_package_test/test_sklearn.py#L117-L118

File: R-package/tests/testthat/test_learning_to_rank.R (hunk 5 above)
Comment (jameslamb) [reply R-package/tests/testthat/test_learning_to_rank.R]: That was my intention, but it's just an artifact of copying other code from the examples you shared. I "one big group" of 513 is probably unnecessary...I'm going to drop this so we keep only the first 6000 observations and split them into 1500 groups of 4, that's cleaner.

File: R-package/tests/testthat/test_learning_to_rank.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (jameslamb) [reply R-package/tests/testthat/test_learning_to_rank.R]: oh sure I can go down to that level. I avoided it because I thought it would make the tests too fragile (even with setting the seed), but if it's been working on the Python side then I'm ok with it. Let me try that.

**Comment by jameslamb** [comment]
> You can merge and lint after on a separate PR.

Thanks for the review! I'm going to fix the linting issues right now. I think it's better that we don't get in the habit of skipping linting and coming back to it later.

File: R-package/tests/testthat/test_learning_to_rank.R (hunk 5 above)
Comment (jameslamb) [reply R-package/tests/testthat/test_learning_to_rank.R]: fixed in https://github.com/microsoft/LightGBM/pull/2573/commits/ead17bcc6ddb06fa506071e76ce0bba213101f8c

File: R-package/tests/testthat/test_learning_to_rank.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (jameslamb) [reply R-package/tests/testthat/test_learning_to_rank.R]: thanks @StrikerRUS , I've made these tests much more specific in https://github.com/microsoft/LightGBM/pull/2573/commits/ead17bcc6ddb06fa506071e76ce0bba213101f8c. Could you take a look?

**Comment by jameslamb** [comment]
Thanks for the reviews @Laurae2 and @StrikerRUS . I think we are really close! I've addressed linting and other issues in the tests in this commit: https://github.com/microsoft/LightGBM/pull/2573/commits/ead17bcc6ddb06fa506071e76ce0bba213101f8c

File: R-package/tests/testthat/test_learning_to_rank.R (hunk 5 above)
Comment (jameslamb) [reply R-package/tests/testthat/test_learning_to_rank.R]: fixed in  https://github.com/microsoft/LightGBM/pull/2573/commits/ee037c9a2f32415c7bd7e63a4d1c093687d110e8 (had to rebase to `master` and force push with other R PRs that were just merged)

File: R-package/tests/testthat/test_learning_to_rank.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (jameslamb) [reply R-package/tests/testthat/test_learning_to_rank.R]: fixed in  https://github.com/microsoft/LightGBM/pull/2573/commits/ee037c9a2f32415c7bd7e63a4d1c093687d110e8 (had to rebase to `master` and force push with other R PRs that were just merged)

**Comment by jameslamb** [comment]
The failures in CI on the previous commit were caused by [this interesting behavior](https://github.com/mailund/pmatch/issues/65#issuecomment-454688322), where basically I was using `&&` when I should have been using  `&`,  and the code range fine on my machine and failed with  `r-devel`  releases.

`&&` should only be  used with inputs that are length-1, and I was using it (incorrectly) to compare to vectors. Just pushed a fix and hopefully that will work.

File: R-package/tests/testthat/test_learning_to_rank.R (hunk 4 above)
[the remark this answers is not in the document]
Comment (StrikerRUS) [reply R-package/tests/testthat/test_learning_to_rank.R]: @jameslamb Thanks you! Tests are more reliable now!

**Comment by jameslamb** [comment]
Thanks for the reviews @Laurae2 and @StrikerRUS . I'm happy to get this one merged and fix this issue.
